### PR TITLE
Add setting "strict" to CodePrinter

### PR DESCRIPTION
**Pull request by bjodah**
For the n'th time I'm debugging generated code containing:
```
# Not supported in Python with numpy:
# Derivative
# Derivative
# Derivative
...
```

This PR aims to introduce a printer option "strict" ~~so that those of us who prefer an exception upon codegen (*fail early*) can opt-in to that behavior~~. The default is now `True`.

#### Brief description of what is fixed or changed
`CodePrinter._default_settings` got a new entry: `strict=True`.

#### Other comments
Subclasses of `CodePrinter` now "inherits" the entries in `CodePrinter._default_settings`. This way, in the future, printer options may be introduced in `codeprinter.py` without touching the respective language's defaults.

#### Release Notes

* printing
  * CodePrinter got a new option "strict" (default is generally "True") to raise an exception, instead of generating comments, upon encountering unsupported constructs in a target language.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
> This PR aims to introduce a printer option "strict" so that those of us who prefer an exception upon codegen (_fail early_) can opt-in to that behavior.

Doesn't everyone want this?

Would it make sense for the default to be strict=True?

**Comment by bjodah** [comment]
I guess you are right. Unless there are those who do some manual postprocessing of the generated code? But I suppose those are few, and we could include some information in the exception message that there is a printer option called "strict" that can be set to False in order to generate partially complete code?

**Comment by bjodah** [comment]
I pushed 18ecb42 to change the default of `strict` to `True`. Or rather, as highlighted by this test:
https://github.com/sympy/sympy/blob/69d3af720c7449a9dcddf17b7e2a4e8724caf15a/sympy/printing/tests/test_fortran.py#L43

The default shouldprobably  only be `True` when `human == True`. So the default is actually `None` and then the value of `human` is used for `strict` if not given explicitly.

**Comment by bjodah** [comment]
So changing the default to `strict=True` was a bit more invasive than anticipated. But now tests pass at least.

File: sympy/printing/tests/test_codeprinter.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def emptyPrinter(self, obj):
     c = CrashingCodePrinter()
 
     # these should not silently succeed
-    with raises(NotImplementedError):
+    with raises(ValueError):
         c.doprint(ImmutableSparseMatrix(2, 2, {}))
-    with raises(NotImplementedError):
+    with raises(ValueError):
         c.doprint(MutableSparseMatrix(2, 2, {}))
```
Comment (oscarbenjamin) [inline sympy/printing/tests/test_codeprinter.py:55]: Would NotImplementedError be more appropriate here?

Comment (oscarbenjamin) [reply sympy/printing/tests/test_codeprinter.py:55]: Or perhaps actually a special exception class.

Comment (bjodah) [reply sympy/printing/tests/test_codeprinter.py]: I agree. I added a new Exception class.

**Comment by oscarbenjamin** [comment]
Looks good!

**Pull request merged**